✨ Refactor context for vspherevm controller

xiujuanx opened the pull request:
<!--  Thanks for sending a pull request! Please add a icon to the title of this PR (see https://sigs.k8s.io/cluster-api-provider-vsphere/CONTRIBUTING.md#contributing-a-patch), and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other)
Here are some other tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide/pull-requests.md#the-pull-request-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
5. If this PR changes image versions, please title this PR "Bump <image name> from x.x.x to y.y.y."
-->

**What this PR does / why we need it**:
Refactor context for vSphereVM controller and underlying functions that vSphereVM controller are consuming, this is a part of https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/issues/2295.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes part of #2295

zhanggbj commented:
/ok-to-test

zhanggbj commented on controllers/vspherevm_controller.go:
```diff
@@ -561,7 +561,7 @@ func (r vmReconciler) retrieveVcenterSession(ctx context.Context, vsphereVM *inf
+	cluster, err := clusterutilv1.GetClusterFromMetadata(ctx, r.Client, vsphereVM.ObjectMeta)
 	if err != nil {
 		r.Logger.Info("VsphereVM is missing cluster label or cluster does not exist")
 		return session.GetOrCreate(r.Context,
```
```suggestion
		return session.GetOrCreate(ctx, params)
```

xiujuanx replied:
Done

zhanggbj commented on controllers/vspherevm_controller.go:
```diff
@@ -607,7 +607,7 @@ func (r vmReconciler) fetchClusterModuleInfo(clusterModInput fetchClusterModuleI
 	logger := r.Logger.WithName(machine.Namespace).WithName(machine.Name)
 
 	input := util.FetchObjectInput{
```
We need to refactor this structure and the related functions  in https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/main/pkg/util/fetch_object.go, to remove the `context.Context` which is our final goal, if there's not too complicated in this PR. Otherwise, please leave a TODO in the comment in #2295 
```
type FetchObjectInput struct {
	context.Context //nolint:containedctx
	ctrlclient.Client
	Object ctrlclient.Object
}
```

xiujuanx replied:
This is not complicated, so I remove context from the `FetchObjectInput struct.

zhanggbj commented on pkg/services/govmomi/service.go:
```diff
@@ -54,27 +55,27 @@ type VMService struct{}
 
 	// Before going further, we need the VM's managed object reference.
-	vmRef, err := findVM(ctx)
+	vmRef, err := findVM(vmCtx)
```
```suggestion
	vmRef, err := findVM(ctx, vmCtx)
```
Need to refactor func https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/main/pkg/services/govmomi/util.go#L60

zhanggbj commented on pkg/services/govmomi/service.go:
```diff
@@ -85,29 +86,29 @@ func (vms *VMService) ReconcileVM(ctx *context.VMContext) (vm infrav1.VirtualMac
 
 		// Get the bootstrap data.
-		bootstrapData, format, err := vms.getBootstrapData(ctx)
+		bootstrapData, format, err := vms.getBootstrapData(vmCtx)
```
```suggestion
		bootstrapData, format, err := vms.getBootstrapData(ctx, vmCtx)
```
Also need to refactor `getBootstrapData`

zhanggbj commented on pkg/services/govmomi/service.go:
```diff
@@ -85,29 +86,29 @@ func (vms *VMService) ReconcileVM(ctx *context.VMContext) (vm infrav1.VirtualMac
 
 		// Create the VM.
-		err = createVM(ctx, bootstrapData, format)
+		err = createVM(vmCtx, bootstrapData, format)
```
```suggestion
		err = createVM(ctx, vmCtx, bootstrapData, format)
```
Need to refactor function `createVM`

zhanggbj commented on pkg/services/govmomi/service.go:
```diff
@@ -118,58 +119,58 @@ func (vms *VMService) ReconcileVM(ctx *context.VMContext) (vm infrav1.VirtualMac
 	vm.VMRef = vmRef.String()
 
-	vms.reconcileUUID(vmCtx)
+	vms.reconcileUUID(virtualMachineCtx)
```
```suggestion
	vms.reconcileUUID(ctx, virtualMachineCtx)
```
need to refactor function `reconcileUUID`

Please also check the following vms.reconcileXXX if they need refactoring

zhanggbj commented on pkg/services/govmomi/service.go:
```diff
@@ -178,26 +179,26 @@ func (vms *VMService) ReconcileVM(ctx *context.VMContext) (vm infrav1.VirtualMac
 	// If there is an in-flight task associated with this VM then do not
 	// reconcile the VM until the task is completed.
-	if inFlight, err := reconcileInFlightTask(ctx); err != nil || inFlight {
+	if inFlight, err := reconcileInFlightTask(vmCtx); err != nil || inFlight {
```
```suggestion
	if inFlight, err := reconcileInFlightTask(ctx, vmCtx); err != nil || inFlight {
```
Need to refactor function reconcileInFlightTask

zhanggbj commented on pkg/services/govmomi/service.go:
```diff
@@ -178,26 +179,26 @@ func (vms *VMService) ReconcileVM(ctx *context.VMContext) (vm infrav1.VirtualMac
 	// there is no task for the VSphereVM resource then no reconcile
 	// event is triggered.
-	defer reconcileVSphereVMOnTaskCompletion(ctx)
+	defer reconcileVSphereVMOnTaskCompletion(vmCtx)
```
same as above

xiujuanx replied:
Done

zhanggbj commented on pkg/services/govmomi/service.go:
```diff
@@ -213,22 +214,22 @@ func (vms *VMService) DestroyVM(ctx *context.VMContext) (reconcile.Result, infra
 
 	// Shut down the VM
-	powerState, err := vms.getPowerState(vmCtx)
+	powerState, err := vms.getPowerState(virtualMachineCtx)
```
```suggestion
	powerState, err := vms.getPowerState(ctx, virtualMachineCtx)
```
Need to refactor function getPowerState

xiujuanx replied:
Done

zhanggbj commented on pkg/services/govmomi/service.go:
```diff
@@ -213,22 +214,22 @@ func (vms *VMService) DestroyVM(ctx *context.VMContext) (reconcile.Result, infra
 	if powerState == infrav1.VirtualMachinePowerStatePoweredOn {
 		// Trigger the soft power off and set the condition.
-		softPowerOffPending, err := vms.triggerSoftPowerOff(vmCtx)
+		softPowerOffPending, err := vms.triggerSoftPowerOff(virtualMachineCtx)
```
same as above

xiujuanx replied:
Done

zhanggbj commented on pkg/services/govmomi/service.go:
```diff
@@ -239,45 +240,45 @@ func (vms *VMService) DestroyVM(ctx *context.VMContext) (reconcile.Result, infra
 
 		// Hard shut off VM.
-		task, err := vmCtx.Obj.PowerOff(ctx)
+		task, err := virtualMachineCtx.Obj.PowerOff(vmCtx)
```
```suggestion
		task, err := virtualMachineCtx.Obj.PowerOff(ctx)
```

zhanggbj commented on pkg/services/govmomi/service.go:
```diff
@@ -239,45 +240,45 @@ func (vms *VMService) DestroyVM(ctx *context.VMContext) (reconcile.Result, infra
+	vmCtx.Logger.Info("VM is powered off", "vmref", vmRef.Reference())
+	if vmCtx.ClusterModuleInfo != nil {
+		provider := clustermodules.NewProvider(vmCtx.Session.TagManager.Client)
+		err := provider.RemoveMoRefFromModule(vmCtx, *vmCtx.ClusterModuleInfo, virtualMachineCtx.Ref)
```
```suggestion
		err := provider.RemoveMoRefFromModule(ctx, *vmCtx.ClusterModuleInfo, virtualMachineCtx.Ref)
```

zhanggbj reviewed (commented):
Thank you for the PR!! 

This is a large one :-) As vspherevm_controller.go will call lots of util functions, many of them need a refactoring. Our goal is to remove the wrapped context.Context in the customized Context [here](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/main/pkg/context/controller_manager_context.go#L35), so any function that's using customized context as a "context", we need to refactor and and a new parameter, that is the right context we need to pass down.

xiujuanx commented on controllers/vspherevm_controller.go:
```diff
@@ -156,11 +155,9 @@ type vmReconciler struct {
 
 // Reconcile ensures the back-end state reflects the Kubernetes resource state intent.
 func (r vmReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
-	r.Logger.V(4).Info("Starting Reconcile", "key", req.NamespacedName)
```
Yeah, I remove it accidentally, add it back now!

xiujuanx commented on controllers/vspherevm_controller.go:
```diff
@@ -221,7 +218,7 @@ func (r vmReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.R
 		}
 
 		vsphereFailureDomain = &infrav1.VSphereFailureDomain{}
-		if err := r.Client.Get(r, apitypes.NamespacedName{Name: vsphereDeploymentZone.Spec.FailureDomain}, vsphereFailureDomain); err != nil {
```
Done!

xiujuanx commented on controllers/vspherevm_controller.go:
```diff
@@ -400,7 +400,7 @@ func (r vmReconciler) reconcileNormal(vmCtx *capvcontext.VMContext) (reconcile.R
 	}
```
Done

zhanggbj commented on pkg/services/govmomi/vcenter/clone.go:
```diff
@@ -338,7 +339,7 @@ func getDiskLocators(disks object.VirtualDeviceList, datastoreRef types.ManagedO
 }
 
-func getDiskSpec(ctx *context.VMContext, devices object.VirtualDeviceList) ([]types.BaseVirtualDeviceConfigSpec, error) {
+func getDiskSpec(_ context.Context, vmCtx *capvcontext.VMContext, devices object.VirtualDeviceList) ([]types.BaseVirtualDeviceConfigSpec, error) {
```
if there's no context or logger using in getDiskSpec, there's no need to add new parameter here.

xiujuanx replied:
Removed the context for `getDiskSpec` func.

zhanggbj commented:
Please address the comments above, then I think we are good to go :-)

xiujuanx commented on controllers/vspherevm_controller.go:
```diff
@@ -462,7 +463,7 @@ func (r vmReconciler) isWaitingForStaticIPAllocation(vmCtx *capvcontext.VMContex
 }
 
-func (r vmReconciler) reconcileNetwork(vmCtx *capvcontext.VMContext, vm infrav1.VirtualMachine) {
+func (r vmReconciler) reconcileNetwork(_ context.Context, vmCtx *capvcontext.VMContext, vm infrav1.VirtualMachine) {
```
Hi @Ankitasw @zhanggbj , I put a new context parameter here is because I am wondering if we want to use the context for some purposes in the future, although there's no context currently for `reconcileNetwork` func. But I check this func again, I find there is no any log or error can be added later, so I agree with you all. I will remove it. Thanks!

xiujuanx commented on controllers/vspherevm_controller.go:
```diff
@@ -260,11 +258,14 @@ func (r vmReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.R
-			reterr = kerrors.NewAggregate([]error{reterr, err})
+			if reterr == nil {
+				reterr = err
+			}
```
I don't intend to modify this part, I think this is caused by I don't rebase the main branch in time.

xiujuanx commented on controllers/vspherevm_ipaddress_reconciler_test.go:
```diff
@@ -47,6 +47,7 @@ func Test_vmReconciler_reconcileIPAddressClaims(t *testing.T) {
 			Logger:            logr.Discard(),
 		}
 	}
+	ctx := context.Background()
```
Reused the pre-defined ctx.

xiujuanx commented on pkg/services/govmomi/cluster/vmgroup.go:
```diff
@@ -26,13 +26,13 @@ import (
 
 // FindVMGroup returns the VSphereVMGroup object.
 // TODO(zhanggbj): add (ctx context.Context) in function signature after refactoring context for pkg/services/govmomi/service.go.
```
Thanks for your kind reminder! I removed this comment, now.

xiujuanx commented on pkg/services/govmomi/template/template.go:
```diff
@@ -35,35 +35,35 @@ type tplContext interface {
 }
 
-func findTemplateByInstanceUUID(ctx tplContext, templateID string) (*object.VirtualMachine, error) {
+func findTemplateByInstanceUUID(ctx context.Context, tplctx tplContext, templateID string) (*object.VirtualMachine, error) {
```
I reuse the pre-defined context, now. Thank you!

xiujuanx commented:
Hi @zhanggbj @Ankitasw , thanks for your review and valuable comments, I am very appreciate for this! Now, I have addressed all comments for this PR, so could you please help take a look again? Thank you again!
BTW, about the golint and apidiff check errors, some of codes I don't touch any more, they are still marked to error. Do I need to fix all errors or not in this PR?

chrischdi commented:
@xiujuanx for golangci-lint: you may have to rebase (just guessing).

apidiff is expected to fail because we change the signature of some functions.

chrischdi commented:
Linter findings seem to be valid though.

This PR is also on my review list :-)

Ankitasw reviewed (commented):
> BTW, about the golint and apidiff check errors, some of codes I don't touch any more, they are still marked to error. Do I need to fix all errors or not in this PR?

@xiujuanx I think it's better we fix those lint errors in this PR itself, otherwise we won't be able to merge this PR, as other PRs would also start breaking with same error. I see majorly the ctx field is unused, we can just mark those fields as empty (_ context.Context). Whenever in further PRs if it is required, we can add back that field declaration.

zhanggbj commented:
@xiujuanx For linter errors, please evaluate them and fix it in this PR, basically if there's logger(which will be refactored in separate PR) or we need context in the function, we should keep the `ctx context.Context` or `_ context.Context` for future usage.

For apidiff check, this is in expectation as we changed signature, and this test is optional for merging :-)

xiujuanx commented:
Thanks @chrischdi @Ankitasw @zhanggbj for your comments and suggestions! I rebased the main branch and fixed some lint errors. Now, the golint job passes.

Ankitasw reviewed (approved):
/lgtm

chrischdi commented on pkg/services/govmomi/util.go:
```diff
@@ -271,39 +272,39 @@ func reconcileVSphereVMOnTaskCompletion(ctx *context.VMContext) {
 
-func reconcileVSphereVMOnChannel(ctx *context.VMContext, waitFn func() (<-chan []interface{}, <-chan error, error)) {
-	obj := ctx.VSphereVM.DeepCopy()
+func reconcileVSphereVMOnChannel(vmCtx *capvcontext.VMContext, waitFn func() (<-chan []interface{}, <-chan error, error)) {
```
We will also need a context.Context here for the logging PR


```suggestion
func reconcileVSphereVMOnChannel(_ context.Context, vmCtx *capvcontext.VMContext, waitFn func() (<-chan []interface{}, <-chan error, error)) {
```

xiujuanx replied:
@chrischdi , I add the `context.Context` for `reconcileVSphereVMOnChannel` func, pls help review it again, thanks!

chrischdi commented:
Only one nit but as of that
/lgtm

zhanggbj commented:
Thanks for the hard word :-)
/lgtm

Ankitasw reviewed (approved):
/lgtm

zhanggbj commented:
/test pull-cluster-api-provider-vsphere-e2e-main

Ankitasw commented:
@zhanggbj looks like the E2E tests are failing due to other reasons. @chrischdi is following up on the same.

chrischdi commented:
/retest

chrischdi commented:
/approve

/hold

if @killianmuldoon wants to take a look too.

chrischdi commented:
/hold cancel
